### ISSUE TITLE
Exception being thrown with Chef 12.6.0

### ISSUE THREAD

**Issue by nickryand** [chef-cookbooks/compat_resource#37]
Method Location: https://github.com/chef-cookbooks/compat_resource/blob/master/files/lib/chef_compat/copied_from_chef/chef/resource_builder.rb#L39

I am running into this really weird issue with the upstream iptables cookbook from Chef. The call to super is throwing an exception when used on ChefDK 10 and Chef 12.6.0. I created a cookbook here that replicates the issue: https://github.com/nickryand/test-iptables.

It feels to me like the conditional operator is returning 'Object' and the 'if defined?' call on line 39 is returning true therefore 'super' is being called. However, since the conditional operator returned 'Object', the constructor does not take parameters. The call to 'super' is passing the parameters from line 38 through and the exception is throw. I do not feel like I fully understand what the code is trying to do so this could be completely wrong.

```
      Recipe: test::default
         * iptables_rule[test] action enable

           ================================================================================
           Error executing action `enable` on resource 'iptables_rule[test]'
           ================================================================================

           ArgumentError
           -------------
           wrong number of arguments (1 for 0)

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/resource_builder.rb:39:in `initialize'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/resource_builder.rb:39:in `initialize'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/dsl/declare_resource.rb:100:in `new'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/dsl/declare_resource.rb:100:in `build_resource'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/dsl/declare_resource.rb:67:in `declare_resource'
           /tmp/kitchen/cache/cookbooks/iptables/resources/rule.rb:26:in `block in class_from_file'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:118:in `instance_eval'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:118:in `compile_and_converge_action'

       Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/test/recipes/default.rb

             9: iptables_rule 'test' do
            10:   action :enable
        11: end

       Compiled Resource:
           ------------------
       # Declared in /tmp/kitchen/cache/cookbooks/test/recipes/default.rb:9:in `from_file'

       iptables_rule("test") do

             retries 0
         retry_delay 2

             declared_type :iptables_rule
             cookbook_name "test"
             recipe_name "default"
           end


       Running handlers:
       [2016-01-13T00:02:34+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2016-01-13T00:02:34+00:00] ERROR: Exception handlers complete
       Chef Client failed. 3 resources updated in 01 seconds
       [2016-01-13T00:02:34+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2016-01-13T00:02:34+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2016-01-13T00:02:34+00:00] ERROR: iptables_rule[test] (test::default line 9) had an error: ArgumentError: wrong number of arguments (1 for 0)
       [2016-01-13T00:02:34+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

**Comment by fishnix** [comment]
We are seeing a very similar issue with the chef-influxdb

```
           ================================================================================
           Error executing action `create` on resource 'influxdb_config[/etc/influxdb/influxdb.conf]'
           ================================================================================

           ArgumentError
           -------------
           wrong number of arguments (1 for 0)

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/resource_builder.rb:39:in `initialize'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/resource_builder.rb:39:in `initialize'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/dsl/declare_resource.rb:100:in `new'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/dsl/declare_resource.rb:100:in `build_resource'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/dsl/declare_resource.rb:67:in `declare_resource'
           /tmp/kitchen/cache/cookbooks/influxdb/resources/config.rb:28:in `block in class_from_file'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:118:in `instance_eval'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:118:in `compile_and_converge_action'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/influxdb/recipes/default.rb

            52: influxdb_config node['influxdb']['config_file_path'] do
            53:   config node['influxdb']['config']
            54: end
            55:
```

**Comment by zhelyan** [comment]
same here with chef 12.5.1

**Comment by bdangit** [comment]
I wonder if we could set a flag or have this lib check current version of the chef_client.  If its 12.5.+ it shouldn't need to do this fancy stuff.

**Comment by jlspies1** [comment]
+1 - am seeing this with iptables cookbook myself.

**Comment by b-dean** [comment]
I found that if I restrict the version of `compat_resource` in my `Berksfile` it works.

```ruby
cookbook 'compat_resource', '= 12.5.10'
```

Obviously that's not a great workaround, but at least it lets me run chef.

**Comment by b-dean** [comment]
I looked at this some more. I believe the problem is that the 
```ruby
defined?(::Chef::ResourceBuilder)
```
for the subclassing is done at the time this file is required. 

The one on line 39 is inside the `initialize method` and will be done later when something is trying to call `Chef::ChefCompat::CopiedFromChef::Chef::ResourceBuilder.new` at which point `::Chef::ResourceBuilder` *will* be defined.

**Comment by b-dean** [comment]
This is still an issue in Chef 12.7.1 and compat_resource 12.7.1
I've rebased my PR that fixes it #41

**Comment by lamont-granquist** [comment]
#45 should have fixed this, open a new issue if not
